File: linked.pas
type
  PNode = ^Node;
  Node = record
    data: Integer;
    next: PNode;
  end;

procedure Create(var head: PNode; value: Integer);
var n: PNode;
begin
  New(n);
  n^.data := value;
  n^.next := head;
  head := n;
end;

procedure Read(head: PNode);
begin
  while head <> nil do
  begin
    Write(head^.data, ' -> ');
    head := head^.next;
  end;
  Writeln('nil');
end;

procedure Delete(var head: PNode; value: Integer);
var temp, prev: PNode;
begin
  temp := head;
  prev := nil;
  while temp <> nil do
  begin
    if temp^.data = value then
    begin
      if prev = nil then head := temp^.next
      else prev^.next := temp^.next;
      Dispose(temp);
      Exit;
    end;
    prev := temp;
    temp := temp^.next;
  end;
end;

File: C/linked.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node *next;
};

struct Node* create(struct Node* head, int data) {
    struct Node* n = malloc(sizeof(struct Node));
    n->data = data;
    n->next = head;
    return n;
}

void read(struct Node* head) {
    while (head) {
        printf("%d -> ", head->data);
        head = head->next;
    }
    printf("NULL\n");
}

struct Node* delete(struct Node* head, int key) {
    struct Node *temp = head, *prev = NULL;
    while (temp) {
        if (temp->data == key) {
            if (prev) prev->next = temp->next;
            else head = temp->next;
            free(temp);
            return head;
        }
        prev = temp;
        temp = temp->next;
    }
    return head;
}
